src/ratingservice/main.py

```diff
@@ -0,0 +1,115 @@
+    db_connection_name = os.environ.get('CLOUD_SQL_CONNECTION_NAME')
+    host = '/cloudsql/{}'.format(db_connection_name)
+    db_config = {
+        'user': db_user,
+        'password': db_password,
+        'database': db_name,
+        'host': host
+    }
+    connpool = pool.ThreadedConnectionPool(minconn=1, maxconn=10, **db_config)
+    return connpool
+
+# read product ids
+def read_products():
+    res = []
+    with open('products.json') as f:
+        data = json.load(f)
+        for product in data['products']:
+            res.append(product['id'])
+    return res
+
+# create and populate table
+def populate_database():
+    try:
+        products = read_products()
+        conn = connpool.getconn()
+        with conn.cursor() as cursor:
+            cursor.execute("SELECT EXISTS(SELECT * FROM information_schema.tables WHERE table_name = 'ratings');")
+            result = cursor.fetchone()
+            if not result[0]:
+                cursor.execute("CREATE TABLE ratings (id SERIAL PRIMARY KEY, product_id varchar(20) NOT NULL, score int DEFAULT 0);")
+                for product in products:
+                    cursor.execute("INSERT INTO ratings (product_id, score) VALUES ('{}', 5);".format(product))
+        conn.commit()
+    finally:
+        connpool.putconn(conn)
+
+app = Flask(__name__)
+connpool = initConnection()
+populate_database()
+
+# get rating of a product
+@app.route('/rating/<id>', methods=['GET'])
+def getRating(id):
+    conn = connpool.getconn()
+    resp = None
+    try:
+        with conn.cursor() as cursor:
+            cursor.execute("SELECT AVG(score), COUNT(*) FROM ratings WHERE product_id='{}';".format(id))
+            result = cursor.fetchone()
+            rating, count = result[0], result[1]
+            if count > 0:
+                # product exists
+                resp = jsonify({
+                    'status' : 'success',
+                    'rating' : str(rating),
+                    'count'  : str(count)
+                })
+                resp.status_code = 200
+            else:
+                # product not exists
+                resp = jsonify({'error' : 'Product not found'.format(id)})
+                resp.status_code = 404   
+        conn.commit()    
+    except:
+        resp = jsonify({'error' : 'Unexpected error in database'})
+        resp.status_code = 500
+    finally:
+        connpool.putconn(conn)
+    return resp
+
+# rate a product
+@app.route('/rating', methods=['POST'])
+def rate():
+    conn = connpool.getconn()
+    resp = None
+    product_id = request.form['id']
+    score = request.form['score']
+    try:
+        with conn.cursor() as cursor:
+            cursor.execute("INSERT INTO ratings (product_id, score) VALUES ('{0}', {1});".format(product_id, score))
+            resp = jsonify({'status' : 'success'})
+            resp.status_code = 200
+        conn.commit()
+    except:
+        resp = jsonify({'error' : 'Unexpected error in database'})
```

Comment:
This error is not descriptive. It's better to say something about the operation that failed. At the very least - "failed to add a rating for XXX" where XXX is the product id

Reply:
Reply to - what is the purpose of the id field: This is because we insert a row at every user rating. If we set product_id to be the primary key, it won't allow duplicate, so there can be only one rating for a product.

Reply to - error not descriptive: I changed it.